agent/xds/clusters.go

```diff
@@ -631,11 +633,51 @@ func (s *Server) makeGatewayCluster(snap *proxycfg.ConfigSnapshot, opts gatewayC
 	}
 	cluster.ClusterDiscoveryType = &discoveryType
 
-	endpoints := make([]envoyendpoint.LbEndpoint, 0, len(opts.hostnameEndpoints))
+	endpoints := make([]envoyendpoint.LbEndpoint, 0, 1)
+	uniqueHostnames := make(map[string]bool)
 
-	for _, e := range opts.hostnameEndpoints {
-		endpoints = append(endpoints, makeLbEndpoint(e, opts.isRemote, opts.onlyPassing))
+	var (
+		hostname string
+		idx      int
+	)
+	for i, e := range opts.hostnameEndpoints {
+		addr, port := e.BestAddress(opts.isRemote)
+		uniqueHostnames[addr] = true
+
+		health, weight := calculateEndpointHealthAndWeight(e, opts.onlyPassing)
+		if health == envoycore.HealthStatus_UNHEALTHY {
+			continue
+		}
+
+		if len(endpoints) == 0 {
+			endpoints = append(endpoints, makeLbEndpoint(addr, port, health, weight))
+
+			hostname = addr
+			idx = i
```

Comment:
```suggestion
			dc := opts.hostnameEndpoints[i].Node.Datacenter
			service := opts.hostnameEndpoints[i].Service.CompoundServiceName()
			break
```

I think you can just break out of the loop when you find the first healthy endpoint, and don't need to save off the index. Instead, you can just get the dc and service here.

Reply:
An issue there is that we wouldn't get to that block if there are no healthy endpoints.

I added a break though.

Reply:
Ah I see, if nothing is healthy you still want to log the dc and service. 👍